File: SplayPrototype.py
class Node:
    def __init__(self, payload, left, right, parent):
        self.payload = payload
        self.left = left
        self.right = right
        self.parent = parent

    def show(self):
        if self.payload is None:
            return "None"
        else:
            return f"({self.payload}, {self.left.show()}, {self.right.show()})"

    def insert(self, payload):
        while self.payload is not None:
            if payload == self.payload:
                return self
            elif payload < self.payload:
                self = self.left
            else:
                self = self.right
        self.payload = payload
        self.left = Node(None, None, None, self)
        self.right = Node(None, None, None, self)
        self.__splay()
        return self

    def __splay(self):
        while self.parent is not None:
            if self.parent.left == self: # left child
                # our right child becomes left child of parent
                self.right.parent = self.parent
                self.parent.left = self.right
                if self.parent.parent is None:
                    # left zig: /
                    parent_in_spe = None
                else:
                    # we become child of great-grandparent
                    parent_in_spe = self.parent.parent.parent
                    if parent_in_spe is not None:
                        if parent_in_spe.left == self.parent.parent:
                            parent_in_spe.left = self
                        else:
                            parent_in_spe.right = self
                    if self.parent.parent.left == self.parent: 
                        # left zig-zig:  /
                        #               /
                        # right child of parent
                        # becomes left child of grandparent
                        self.parent.right.parent = self.parent.parent
                        self.parent.parent.left = self.parent.right
                        # grandparent becomes right child of parent
                        self.parent.parent.parent = self.parent
                        self.parent.right = self.parent.parent
                    else:                         
                        # right-left zig-zag: \
                        #                     /
                        # our left child becomes right child of grandparent
                        self.left.parent = self.parent.parent
                        self.parent.parent.right = self.left
                        # grandparent becomes our left child
                        self.parent.parent.parent = self
                        self.left = self.parent.parent
                # parent becomes our right child
                self.parent.parent = self
                self.right = self.parent
                # great-grandparent that we had to save becomes our parent
                self.parent = parent_in_spe
            else: # right child
                # our left child becomes right child of parent
                self.left.parent = self.parent
                self.parent.right = self.left
                if self.parent.parent is None:
                    # right zig: \
                    parent_in_spe = None
                else:
                    # we become child of great-grandparent
                    parent_in_spe = self.parent.parent.parent
                    if parent_in_spe is not None:
                        if parent_in_spe.left == self.parent.parent:
                            parent_in_spe.left = self
                        else:
                            parent_in_spe.right = self
                    if self.parent.parent.right == self.parent: 
                        # right zig-zig: \
                        #                 \
                        # left child of parent
                        # becomes right child of grandparent
                        self.parent.left.parent = self.parent.parent
                        self.parent.parent.right = self.parent.left
                        # grandparent becomes left child of parent
                        self.parent.parent.parent = self.parent
                        self.parent.left = self.parent.parent
                    else:                         
                        # left-right zig-zag: /
                        #                     \
                        # our right child becomes left child of grandparent
                        self.right.parent = self.parent.parent
                        self.parent.parent.left = self.right
                        # grandparent becomes our right child
                        self.parent.parent.parent = self
                        self.right = self.parent.parent
                # parent becomes our left child
                self.parent.parent = self
                self.left = self.parent
                # great-grandparent that we had to save becomes our parent
                self.parent = parent_in_spe


class SplayTree:
    def __init__(self):
        self.tree = Node(None, None, None, None)

    def show(self):
        return self.tree.show()

    def insert(self, payload):
        self.tree = self.tree.insert(payload)

File: CPrototype/SplayPrototype.c
#include <stdio.h>
#include <stdlib.h>
#include "Enums.h"
#include "Node.h"
#include "NodeStack.h"
#include "BoolStack.h"

struct node* initTree() {
  struct node* node = malloc(sizeof(struct node));
  node->payload = NULL;
  node->left = NULL;
  node->right = NULL;
  return node;
}

void clearTree(struct node* node) {
  if (node->left != NULL) clearTree(node->left);
  if (node->right != NULL) clearTree(node->right);
  free(node);
}

// internal function that does not print a newline
void showTreeRec(struct node* node, void (*show)(void*)) {
  if (node == NULL || node->payload == NULL)
    printf("NULL");
  else {
    printf("(");
    (*show)(node->payload);
    printf(", ");
    showTreeRec(node->left, show);
    printf(", ");
    showTreeRec(node->right, show);
    printf(")");
  }
}

void showTree(struct node* node, void (*show)(void*)) {
  showTreeRec(node, show);
  printf("\n");
}

void splay(struct node* node, struct nodeStack* parentStack, struct boolStack* directionStack) {
  while (isEmptyNode(parentStack) == FALSE) {
    struct node* parent = popNode(parentStack);
    enum boolean isLeftChild = popBool(directionStack);
    if (isEmptyNode(parentStack) == TRUE) {
      // zig
      if (isLeftChild == TRUE) {
        /* left zig: / */
        parent->left = node->right;
        node->right = parent;
      } else {
        /* right zig: \ */
        parent->right = node->left;
        node->left = parent;
      }
    } else {
      // zig-zig or zig-zag
      struct node* grandparent = popNode(parentStack);
      enum boolean isLeftGrandchild = popBool(directionStack);
      if (isLeftChild == TRUE && isLeftGrandchild == TRUE) {
        /* left zig-zig:  /
         *               /  */
        grandparent->left = parent->right;
        parent->left = node->right;
        parent->right = grandparent;
        node->right = parent;
      } else if (isLeftChild == FALSE && isLeftGrandchild == TRUE) {
        /* left-right zig-zag: /
         *                     \ */
        parent->right = node->left;
        grandparent->left = node->right;
        node->left = parent;
        node->right = grandparent;
      } else if (isLeftChild == FALSE && isLeftGrandchild == FALSE) {
        /* right zig-zig: \
         *                 \ */
        grandparent->right = parent->left;
        parent->left = grandparent;
        parent->right = node->left;
        node->left = parent;
      } else {
        /* right-left zig-zag: \
         *                     / */
        grandparent->right = node->left;
        parent->left = node->right;
        node->left = grandparent;
        node->right = parent;
      }
    }
  }
  free(parentStack);
  free(directionStack);
}

struct node* insert(struct node* node, void* payload, enum order (*order)(void*, void*)) {
  struct nodeStack* parentStack = initNodeStack();
  struct boolStack* directionStack = initBoolStack();
  while (1) {
    // break using return when inserted
    if (node->payload == NULL) {
      // empty tree
      node->payload = payload;
      return node;
    } else if ((*order)(payload, node->payload) == LESS) {
      if (node->left == NULL) {
        struct node* leaf = malloc(sizeof(struct node));
        leaf->payload = payload;
        leaf->left = NULL;
        leaf->right = NULL;
        node->left = leaf;
        /* splay new child up, appending self to the parent stack,
         * the child being a left child of self */
        pushNode(parentStack, node);
        pushBool(directionStack, TRUE);
        splay(leaf, parentStack, directionStack);
        // newly inserted node is new root after splay
        return leaf;
      } else {
        // insert to the left recursively, child is left child
        pushNode(parentStack, node);
        pushBool(directionStack, TRUE);
        node = node->left;
      }
    } else if ((*order)(payload, node->payload) == EQUALS) {
      // element is present (no double elements in sets), just splay up
      splay(node, parentStack, directionStack);
      // node is root after splay
      return node;
    } else {
      if (node->right == NULL) {
        struct node* leaf = malloc(sizeof(struct node));
        leaf->payload = payload;
        leaf->left = NULL;
        leaf->right = NULL;
        node->right = leaf;
        // splay (right) child up
        pushNode(parentStack, node);
        pushBool(directionStack, FALSE);
        splay(leaf, parentStack, directionStack);
        return leaf;
      } else {
        // insert on right child recursively
        pushNode(parentStack, node);
        pushBool(directionStack, FALSE);
        node = node->right;
      }
    }
  }
}

